### PR TITLE
[`pyupgrade`] Rename private type parameters in PEP 695 generics (`UP049`)

### PR DESCRIPTION
**Pull request by ntBre**
## Summary

This is a new rule to implement the renaming of PEP 695 type parameters with leading underscores after they have (presumably) been converted from standalone type variables by either UP046 or UP047. Part of #15642.

I'm not 100% sure the fix is always safe, but I haven't come up with any counterexamples yet. `Renamer` seems pretty precise, so I don't think the usual issues with comments apply.

I initially tried writing this as a rule that receives a `Stmt` rather than a `Binding`, but in that case the `checker.semantic().current_scope()` was the global scope, rather than the scope of the type parameters as I needed. Most of the other rules using `Renamer` also used `Binding`s, but it does have the downside of offering separate diagnostics for each parameter to rename.

## Test Plan

New snapshot tests for UP049 alone and the combination of UP046, UP049, and PYI018.

### CODE REVIEW

File: crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    y: _T = var[1]
+    return y
+
+
+# one diagnostic for each variable, comments are preserved
+def many_generics[
+    _T,  # first generic
+    _U,  # second generic
+](args):
+    return args
```
[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: No the `Renamer` does not catch these, I guess that's enough to make the fix unsafe.

[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: Never mind, they *both* get renamed when `cast` is actually imported. Why should the second one not get renamed? Is it supposed to be outside of the function? I "fixed" the indentation by putting them both inside.

[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: Ah right, thanks. I guess I spoke too soon again. If `Literal` is also imported, I think we get the correct behavior for your original example (the first one replaced, the second not). I will throw the `\x54` case in as well just for fun. Thanks for these additional cases!

[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: Well this isn't good. `cast('list[_\x54]', v) ` gets replaced with `cast(T, v)`, while `cast("list[_T]", v)` gets replaced correctly with `cast("list[T]", v)`. I'm not sure why the hex version is deleting the `list[]` part. Is this something you've run into before?

File: crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap:1]: I'm not exactly sure why the UP051 errors aren't being shown in this snapshot. I don't think this indicates a bug in the logic for your rule, though, as everything seems to work correctly if I run the rules manually together on this fixture with your branch:

````
~/dev/ruff (brent/rename-typevars) [1] % cargo run -- check --no-cache --preview --diff --select="PYI018,UP051,UP046,F401" --target-version=py313 --unsafe-fixes crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
     Running `target/debug/ruff check --no-cache --preview --diff --select=PYI018,UP051,UP046,F401 --target-version=py313 --unsafe-fixes crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py`
--- crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py
+++ crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py
@@ -1,4 +1,4 @@
-from typing import Generic, TypeVar
+from typing import TypeVar
 
 # this case should not be affected by UP051 alone, but with UP046 and PYI018,
 # the old-style generics should be replaced with type parameters, the private
@@ -7,5 +7,5 @@
 _T = TypeVar("_T")
 
 
-class OldStyle(Generic[_T]):
-    var: _T
+class OldStyle[T]:
+    var: T

Would fix 3 errors.
````

The only thing that surprises me there is that the PYI018 violation introduced by the UP046 change introduces is not autofixed. Apparently we don't have an autofix for PYI018 right now; I suppose I assumed that we did!

My guess is that the `assert_messages!()` macro we use in our tests maybe doesn't run Ruff in a loop in the same way as `ruff check` does when it's run from the command line? Not sure. @MichaReiser, any ideas?

One solution might be to write this as a CLI integration test, if our standard testing infrastructure isn't up to the task here.

File: crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: > Well this isn't good. `cast('list[_\x54]', v) ` gets replaced with `cast(T, v)`, while `cast("list[_T]", v)` gets replaced correctly with `cast("list[T]", v)`. I'm not sure why the hex version is deleting the `list[]` part. Is this something you've run into before?

This sounds like it could be a bug in the `Renamer`. It also seems like an extreme edge case, however, that's unlikely to actually come up in either handwritten or generated code -- so personally I think the fix could still be marked as safe.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+/// also want to consider enabling [`unused-private-type-var`](unused-private-type-var.md) to
+/// complete the transition to [PEP 695] generics.
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+#[derive(ViolationMetadata)]
+pub(crate) struct PrivateTypeParameter {
+    kind: ParamKind,
+}
+
+enum ParamKind {
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:55]:
```suggestion
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum ParamKind {
```

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            ..
+        }) => (ParamKind::Function, type_params.range),
+        Stmt::ClassDef(StmtClassDef {
+            type_params: Some(type_params),
+            ..
+        }) => (ParamKind::Class, type_params.range),
+        _ => return None,
+    };
+
+    let old_name = binding.name(checker.source());
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: I came across this the other day -- unfortunately I think `binding.name()` does the wrong thing here for PEP-695 type parameters with bounds or constraints. E.g. for `_T` here:

```py
class Foo[_T: str]: ...
```

I believe `binding.name()` will return `"_T: str"` rather than `"_T"`

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    if !old_name.starts_with('_') {
+        return None;
+    }
+
+    let new_name = old_name.trim_start_matches('_');
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:100]: What happens here if we have a PEP-695 variable named `_` or `__`?

I think this will also flag e.g. `class Foo[_T_]: ...` and `class Foo[__T__]: ...` ("sunder" and "dunder" variables, which are arguably not "private"). Do we want to skip the rule if the name also ends in `_`?

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    let mut diagnostic = Diagnostic::new(PrivateTypeParameter { kind }, range);
+
+    diagnostic.try_set_fix(|| {
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:104]: we probably want to check various thinsg before doing the rename, such as:
- whether the new name would be a keyword (changing `class F[_async]: ...` to `class F[async]: ...` would introduce invalid syntax)
- whether the new name would shadow an existing variable in an outer scope

A lot of this logic is already done in this function for RUF052:

https://github.com/astral-sh/ruff/blob/83243de93de6b63ab6de3a6b77113416c3eb51d9/crates/ruff_linter/src/rules/ruff/rules/used_dummy_variable.rs#L239-L271

We could possibly move that function to `renamer.rs` if it would also be helpful here

**Review by AlexWaygood — COMMENTED** [review]
Overall this looks great!

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+/// ```
+///
+/// ## See also
+///
+/// This rule renames private [PEP 695] type parameters but doesn't convert pre-[PEP 695] generics
+/// to the new format. See [`non-pep695-generic-function`](non-pep695-generic-function.md) and
+/// [`non-pep695-generic-class`](non-pep695-generic-class.md) for rules that will make this
+/// transformation. Those rules do not remove unused type variables after their changes, so you may
+/// also want to consider enabling [`unused-private-type-var`](unused-private-type-var.md) to
+/// complete the transition to [PEP 695] generics.
```
[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:47]: Thanks, and I see you already updated the related rules in https://github.com/astral-sh/ruff/pull/15904.

File: crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ntBre) [reply crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051.py:20]: Sounds good. For now I've included the other three cases and left the rule as safe. I'll try to add a `Renamer` test for the hex case and debug that separately.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 4 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:100]: Ah great catches again. I think it makes sense to skip variables that also end in `_` and obviously if the new name is empty :facepalm:

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 3 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: This also completely skips `TypeVarTuple`s and `ParamSpec`s because they "start with" `*` and `**`.

Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: I think I will have to do some work on either `Binding` or `Renamer`. I tried looping over the `type_params` directly to get just the `_T` part of the name, but `Renamer` finds the `Binding` for the name I'm giving it and replaces the whole `Binding::range` anyway.

Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: I think I have the `Binding` range fix. I was considering a separate PR, but it's a one-line change in the semantic model and two one-line changes in PYI019, so I think I'll just include it here. This rule would have to be reverted with it anyway.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 5 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:104]: Nice, thanks for this suggestion! It's kind of a shame to do the `name.starts_with` and `trim_start_matches` calls twice in both the new rule here and in `used_dummy_variable`. I might try to tweak the API a bit, but the current version passes new tests I added based on your comment!

Do you think we should offer a diagnostic and not a fix here, or bail out entirely? I'm currently bailing out, but it seems fairly reasonable to have a diagnostic, at least without thinking too hard.

File: crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap:1]: I poked around a bit more and from what I can tell, `test_contents` returns the first [`diagnostics`](https://github.com/astral-sh/ruff/blob/brent%2Frename-typevars/crates/ruff_linter/src/test.rs#L123) assignment, even though it does run the fixes in a loop and see the [other](https://github.com/astral-sh/ruff/blob/brent%2Frename-typevars/crates/ruff_linter/src/test.rs#L188) diagnostics. I went ahead and converted to an integration test for now!

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 3 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: > I was considering a separate PR, but it's a one-line change in the semantic model and two one-line changes in PYI019, so I think I'll just include it here.

Hmm, I think I _would_ prefer this as a separate PR for a couple of reasons:
1. I think we can actually simplify the code in PYI019 further with this fix
2. Changing the range of `Binding` makes me a _little_ nervous that we might actually change the range of some diagnostics, which would be a breaking change. I'm _pretty_ confident from the ecosystem report and the fact that no snapshots in other rules need to change that we're not changing any diagnostic ranges... but I'm not _100%_ confident. If we move it to an isolated change, it will be easy to bisect it (if necessary) and see exactly what caused the breaking change.

I know exactly what I'd like to do for point (1) here, so I'll file a new PR and list you as a co-author -- hope that's okay!

Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: That makes sense, it definitely makes me nervous too! I was surprised only one test changed as a result.

Sounds good about the the separate PR!

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: https://github.com/astral-sh/ruff/pull/15935

Sorry for the delay, got distracted by things I thought would take a little less time than they did!

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 5 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:110]: I think having a diagnostic (but not offering a fix) makes sense!

RUF052 also does some stuff to try to find other names it could use instead if the name-sans-leading-underscore is shadowed/keyword/not-an-identifier. I don't think it makes sense to do the same here, though; I think we should just have a diagnostic-without-fix if it's shadowed/keyword/not-an-identifier

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 3 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: No worries at all! I'll merge main again.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 5 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:110]: I agree, I'll update this and add a test showing just a diagnostic.

Are we safe to handle `default` here? With the `Binding` change we're just replacing the name, so I was thinking about adding a test with a `default` type parameter too.

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:110]: > Are we safe to handle `default` here? With the `Binding` change we're just replacing the name, so I was thinking about adding a test with a `default` type parameter too.

Yeah, I don't think TypeVars with defaults need any special treatment here! We should be good to handle those in this rule 👍

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 3 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: ...and also this one 😆

- https://github.com/astral-sh/ruff/pull/15938

Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:95]: Ahh I thought I had to add `.name()` in two places! Good catch!

File: crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap:1]: Should we open an issue to add a PYI018 autofix? It might be a good `help-wanted` or `first-issue`, unless I'm underestimating it of course.

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP051_1.py.snap:1]: that would be great, yeah. I think it would be very useful

File: crates/ruff/tests/snapshots/lint__pep695_generic_rename.snap
Comment (AlexWaygood) [inline crates/ruff/tests/snapshots/lint__pep695_generic_rename.snap:1]: nit: I find it much easier to read these integration tests if we use inline snapshots, like this test does:

https://github.com/astral-sh/ruff/blob/5bf0e2e95ed4c9ef3e063c6e3876c59d827cb9cf/crates/ruff/tests/lint.rs#L22-L63

(`cargo insta` can still update the snapshot contents interactively, just like with separate-file snapshots)

File: crates/ruff_linter/src/renamer.rs [hunk 7]
```diff
@@ -369,3 +372,69 @@ impl Renamer {
...
+    checker: &Checker,
+    scope_id: ScopeId,
+) -> Option<ShadowedKind> {
+    // If the name doesn't start with an underscore, we don't consider it
+    if !name.starts_with('_') {
+        return None;
+    }
+
+    // Trim the leading underscores for further checks
+    let trimmed_name = name.trim_start_matches('_');
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/renamer.rs:416]: This feels too specific to RUF052 and UP051 to be included in the generic utility in this module (not all potential users of the `Renamer` will be renaming names that start with underscores). I think we should be checking whether it starts with an underscore (and trimming leading underscores) in RUF052/UP051, and then passing the trimmed name into `try_shadowed_kind()`. This would also allow `try_shadowed_kind()` to return `ShadowedKind` rather than `Option<ShadowedKind>`.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+///
+///
+/// def generic_function[T](var: T) -> list[T]:
+///     return var[0]
+/// ```
+///
+/// ## Fix availability
+///
+/// This rule avoids shadowing builtins, keywords, and other variables with new type parameter names
+/// and will offer a diagnostic but not a fix in the case of a conflict.
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:46]:
```suggestion
/// If the name without an underscore would shadow a builtin or another variable,
/// would be a keyword, or would otherwise be an invalid identifier, a fix will not
/// be available. In these situations, you can consider using a trailing underscore
/// or a different name entirely to satisfy the lint rule.
```

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+/// This rule renames private [PEP 695] type parameters but doesn't convert pre-[PEP 695] generics
+/// to the new format. See [`non-pep695-generic-function`] and [`non-pep695-generic-class`] for
+/// rules that will make this transformation. Those rules do not remove unused type variables after
+/// their changes, so you may also want to consider enabling [`unused-private-type-var`] to complete
+/// the transition to [PEP 695] generics.
+///
+/// [PEP 695]: https://peps.python.org/pep-0695/
+/// [`non-pep695-generic-function`]: https://docs.astral.sh/ruff/rules/non-pep695-generic-function
+/// [`non-pep695-generic-class`]: https://docs.astral.sh/ruff/rules/non-pep695-generic-class
+/// [`unused-private-type-var`]: https://docs.astral.sh/ruff/rules/unused-private-type-var
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:59]: I can never remember whether mkdocs will create links correctly if backticks are included in the `[]` names for the footnotes...

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+        _ => return None,
+    };
+
+    let old_name = binding.name(checker.source());
+
+    // this covers the sunder `_T_`, dunder `__T__`, and all all-under `_` or `__` cases
+    // that should be skipped
+    if !old_name.starts_with('_') || old_name.ends_with('_') {
+        return None;
+    }
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:107]: nit: I'd find this a bit more readable; these feel like different conditions

```suggestion
    if !old_name.starts_with('_') {
        return None;
    }
     
    // Sunder `_T_`, dunder `__T__`, and all all-under `_` or `__` cases
    // should all be skipped, as these are not "private" names
    if old_name.ends_with('_') {
        return None;
    }
```

It would also be good to mention this logic explicitly in the docs for the rule!

File: crates/ruff_linter/src/renamer.rs [hunk 11]
```diff
@@ -369,3 +372,69 @@ impl Renamer {
...
+    BuiltIn,
+    /// The variable shadows a keyword
+    Keyword,
+    /// The variable does not shadow any other symbols
+    None,
+}
+
+impl ShadowedKind {
+    /// Returns `true` if `self` shadows any global, nonlocal, or local symbol, keyword, or builtin.
+    pub(crate) fn shadows_any(self) -> bool {
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/renamer.rs:391]:
```suggestion
    pub(crate) const fn shadows_any(self) -> bool {
```

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+impl Violation for PrivateTypeParameter {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let kind = match self.kind {
+            ParamKind::Class => "class",
+            ParamKind::Function => "function",
+        };
+        format!("Generic {kind} uses private type parameters")
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:79]:
```suggestion
}

impl ParamKind {
    const fn as_str(self) -> &'static str {
        match self {
            Self::Class => "class",
            Self::Function => "function",
        }
    }
}

impl Violation for PrivateTypeParameter {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
    #[derive_message_formats]
    fn message(&self) -> String {
        format!("Generic {} uses private type parameters", self.kind.as_str())
```

File: crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+
+# and this should not be fixed because of the conflict with the outer X, but it
+# also gets a diagnostic
+def f():
+    X = 5
+
+    class ScopeConflict[_X]:
+        var: _X
+        x: X
```
Comment (AlexWaygood) [inline crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py:49]: it might be more realistic if `X` is an actual type that it would be valid to annotate `x` with?

```suggestion
    type X = int

    class ScopeConflict[_X]:
        var: _X
        x: X
```

File: crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py
Comment (AlexWaygood) [inline crates/ruff_linter/resources/test/fixtures/pyupgrade/UP051_1.py:1]: Could you possibly add these tests, just for completeness? I think none of them should cause us to emit diagnostics:

```py
def f[_](x: _) -> _: ...
def g[__](x: __) -> __: ...
def h[_T_](x: _T_) -> _T_: ...
def i[__T__](x: __T__) -> __T__: ...
```

File: crates/ruff/tests/snapshots/lint__pep695_generic_rename.snap
Comment (ntBre) [reply crates/ruff/tests/snapshots/lint__pep695_generic_rename.snap:1]: Good call, I copied my (poor) example from the test above.

File: crates/ruff_linter/src/renamer.rs (hunk 7 above)
Comment (ntBre) [reply crates/ruff_linter/src/renamer.rs:416]: oh good idea! I also moved `try_shadowed_kind` to an associated function `ShadowedKind::new`. Definitely open to a better name there or reverting to a standalone function!

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 9 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:59]: Oh good to know. The S606 example linked above does not have the quotes and renders correctly on the website, so I'll remove these to be safe.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 12 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:79]: Do you ever make the `Violation` type an enum itself? I just tried that locally and it seems to be working:

```rust
#[derive(ViolationMetadata, Debug, Copy, Clone, PartialEq, Eq)]
pub(crate) enum PrivateTypeParameter {
    Class,
    Function,
}
```

but I'm not sure if that will cause problems somewhere.

I'll just go with your suggestion here.

Comment (AlexWaygood) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:79]: hmm, dunno! That could work! @MichaReiser redesigned our `ViolationMetadata` macro recently; it's possible that it wasn't possible to do that before, but it now is. Not sure!

File: crates/ruff_linter/src/rules/ruff/rules/used_dummy_variable.rs [hunk 14]
```diff
@@ -164,53 +165,48 @@ pub(crate) fn used_dummy_variable(
...
-    if let Some(shadowed_kind) = shadowed_kind {
-        // Get the possible fix based on the scope
-        if let Some(fix) = get_possible_fix(name, shadowed_kind, binding.scope, checker) {
-            diagnostic.try_set_fix(|| {
-                Renamer::rename(name, &fix, scope, semantic, checker.stylist())
-                    .map(|(edit, rest)| Fix::unsafe_edits(edit, rest))
-            });
-        }
+    // Get the possible fix based on the scope
+    if let Some(fix) = get_possible_fix(trimmed_name, shadowed_kind, binding.scope, checker) {
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/ruff/rules/used_dummy_variable.rs:193]: I know it's not an issue that you're introducing here, but I'd expect a function with the name `get_possible_fix` to return a `Fix` rather than `Option<String>`. Maybe we could rename it to `get_possible_new_name` or similar?

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+///
+/// ## Fix availability
+///
+/// If the name without an underscore would shadow a builtin or another variable, would be a
+/// keyword, or would otherwise be an invalid identifier, a fix will not be available. In these
+/// situations, you can consider using a trailing underscore or a different name entirely to satisfy
+/// the lint rule.
+///
+/// Additionally, neither a diagnostic nor a fix will be emitted for "sunder" (`_T_`) or "dunder"
+/// (`__T__`) type parameter names as these are not considered private.
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:51]: I would maybe move this up and make it a second paragraph in the "Why is this bad?" section, since it's also relevant for users who don't opt into autofixes

**Comment by ntBre** [comment]
Awesome, thanks for all of your help! Is UP051 an acceptable code to use? I just searched issues and PRs again, and the highest number I see now is UP048 (#15625). I think #15841 was previously UP050, which is probably where I got UP051. I'm happy to switch to UP049 to avoid leaving a gap, if you want.

**Comment by AlexWaygood** [comment]
It's nice to avoid leaving a gap where possible with the rule codes, but I don't think it matters too much in cases like this where there's no corresponding "upstream" rule! It _is_ quite important to make sure we don't use the same code as a rule that we used to have but has since been deprecated or removed, however.

File: crates/ruff_linter/src/rules/ruff/rules/used_dummy_variable.rs (hunk 14 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/ruff/rules/used_dummy_variable.rs:193]: I came up with `get_candidate_name` and `try_get_name` too, but I like `get_possible_new_name`.

File: crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs (hunk 15 above)
Comment (ntBre) [reply crates/ruff_linter/src/rules/pyupgrade/rules/pep695/private_type_parameter.rs:51]: Ah makes sense, I wasn't sure where to put it, but near the top does seem better.

**Review by AlexWaygood — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Oh shoot, I forgot to mention. Can we add a mention of this rule to the `## See also` sections in the docs for UP046, UP047 and UP040? The autofixes for those rules will work really well if this one is enabled as well

**Comment by ntBre** [comment]
Ahhh, good catch. I meant to do that the whole time but forgot. I'll open a follow-up!
